### DISCUSSION TITLE
Application freezes on first voice request

### DISCUSSION THREAD

**Discussion by jeffries7** [RageAgainstThePixel/com.rest.elevenlabs#67] [Q&A] [answered]
I think this is more of a question than a bug but we're seeing large spikes when processing text.

We're using the latest version of the SDK & Unity 2021.3.21f1.
This is running on a Meta Quest 3 but we're also seeing this when running in the editor.

Pretty much all requests we make after the first one are absolutely fine but the first one freezes the application for a second or two.

I've taken a look at the samples but it has the same results.

Below is the code we've been using for a while.
```csharp
    public void ExecuteTask(string answer)
    {
         if (_isReady)
             Execute(answer);
    }
    
    private async void Execute(string answer)
    {
        _isReady = false;

        var api = new ElevenLabsClient();

        var voice = await api.VoicesEndpoint.GetVoiceAsync(voiceData.DefaultVoice.ID);

        var voiceClip = await api.TextToSpeechEndpoint.TextToSpeechAsync(answer, voice, model: Model.EnglishTurboV2);

        if (_audioSource.isPlaying)
        {
            _audioSource.Stop();
        }

        onAudioProcessed?.Invoke();

        _audioSource.clip = voiceClip.AudioClip;
        _audioSource.Play();

        StartCoroutine(CheckAudio());
    }
```

**Comment by StephenHodgson** [comment]
the freeze you're likely seeing is the voice clip being loaded into memory.

you might see better performance with the streaming endpoint.

**Reply by StephenHodgson** [reply]
@jeffries7 if you can get me some profiler info about where specifically you're seeing the freeze we can look into optimizing more.

**Reply by jeffries7** [reply]
I've done some more testing in a blank scene in the editor, I wanted to remove the parameter of it being related to running on the Quest.

This is a capture from the first time calling StreamTextToSpeechAsync. Massive CPU spikes, up to around 90ms.
![image](https://github.com/RageAgainstThePixel/com.rest.elevenlabs/assets/19264723/03bbfe98-da3c-420d-b5cf-bbb97bb897b4)

Every other call after results in tiny usages.
I'm passing a single word into StreamTextToSpeechAsync too.


My temporary fix is to make a small request when the application is started as the dropped frames are hidden within the black loading screen. Then every request after works fine. We're not even using the stream functionality just the regular audio request as the audio quality is higher.

**Reply by StephenHodgson** [reply]
I bet it is the client credential setup phase.

Are you also getting the warnings about perf and passing in the configuration through the constrictor?

How are you passing the API key?

**Reply by jeffries7** [reply]
We're passing the api key via a scriptable object. I assume we should be using a config file instead as I don't think we can use an environment variable on the Quest.

I've just seen we have these warnings.  

![image](https://github.com/RageAgainstThePixel/com.rest.elevenlabs/assets/19264723/33e37eb5-1870-4294-872a-559866146e4f)

**Reply by StephenHodgson** [reply]
Ideally you shouldn't be shipping your API keys in your app anyway.

**Reply by StephenHodgson** [reply, marked as answer]
See [API proxy instructions](https://github.com/RageAgainstThePixel/com.rest.elevenlabs#api-proxy).

And to get around the performance impact create the auth and settings objects before constructing your client.

**Reply by jeffries7** [reply]
Thank you so much for this. Its a 10/10 package, keep up the great work and great support!

**Comment by StephenHodgson** [comment]
By the way you can await this coroutine:

```
await CheckAudio();
```

**Reply by jeffries7** [reply]
You're right, thanks 👍

**Comment by StephenHodgson** [comment]
Duplicate of https://github.com/RageAgainstThePixel/com.rest.elevenlabs/discussions/69
